components/director/internal/domain/runtime/service.go

```diff
@@ -116,48 +140,132 @@ func (s *service) Update(ctx context.Context, id string, in model.RuntimeInput)
 		return errors.Wrap(err, "while updating Runtime")
 	}
 
+	rtmTenant, err := tenant.LoadFromContext(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "while loading tenant from context")
+	}
+
+	err = s.labelRepo.DeleteAll(ctx, rtmTenant, model.RuntimeLabelableObject, id)
+	if err != nil {
+		return errors.Wrapf(err, "while deleting all labels for Runtime")
+	}
+
+	err = s.labelUpsertService.UpsertMultipleLabels(ctx, rtmTenant, model.RuntimeLabelableObject, id, in.Labels)
+	if err != nil {
+		return errors.Wrapf(err, "while creating multiple labels for Runtime")
+	}
+
 	return nil
 }
 
 func (s *service) Delete(ctx context.Context, id string) error {
-	rtm, err := s.Get(ctx, id)
+	rtmTenant, err := tenant.LoadFromContext(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "while loading tenant from context")
+	}
+
+	exists, err := s.repo.Exists(ctx, rtmTenant, id)
 	if err != nil {
 		return errors.Wrap(err, "while getting Runtime")
 	}
+	if !exists {
+		return fmt.Errorf("Runtime %s doesn't exist", id)
+	}
 
-	return s.repo.Delete(ctx, rtm.ID)
+	err = s.repo.Delete(ctx, id)
+	if err != nil {
+		return errors.Wrapf(err, "while deleting Runtime")
+	}
+
+	// All labels are deleted (cascade delete)
+
+	return nil
 }
 
-func (s *service) SetLabel(ctx context.Context, runtimeID string, key string, value interface{}) error {
-	rtm, err := s.Get(ctx, runtimeID)
+func (s *service) SetLabel(ctx context.Context, labelInput *model.LabelInput) error {
+	rtmTenant, err := tenant.LoadFromContext(ctx)
 	if err != nil {
-		return errors.Wrap(err, "while getting Runtime")
+		return errors.Wrapf(err, "while loading tenant from context")
 	}
 
-	rtm.SetLabel(key, value)
+	rtmExists, err := s.repo.Exists(ctx, rtmTenant, labelInput.ObjectID)
+	if err != nil {
+		return errors.Wrap(err, "while checking Runtime existence")
+	}
+	if !rtmExists {
+		return fmt.Errorf("Runtime with ID %s doesn't exist", labelInput.ObjectID)
+	}
 
-	err = s.repo.Update(ctx, rtm)
+	err = s.labelUpsertService.UpsertLabel(ctx, rtmTenant, labelInput)
 	if err != nil {
-		return errors.Wrapf(err, "while updating Runtime")
+		return errors.Wrapf(err, "while creating label for Runtime")
 	}
 
 	return nil
 }
 
+func (s *service) GetLabel(ctx context.Context, runtimeID string, key string) (*model.Label, error) {
+	rtmTenant, err := tenant.LoadFromContext(ctx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while loading tenant from context")
+	}
+
+	rtmExists, err := s.repo.Exists(ctx, rtmTenant, runtimeID)
```

Comment:
do we really need to have this logic in code when we have cascading delete in code?

Reply:
As discussed, I will keep it as it is.